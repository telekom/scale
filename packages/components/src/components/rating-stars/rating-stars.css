/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --stars-color: var(--scl-color-primary);
  --stars-hover-color: var(--scl-color-primary-hover);
  --stars-active-color: var(--scl-color-primary-active);
  --stars-inactive-color: transparent;
  --stars-placeholder-color: var(--scl-color-black);
  --stars-placeholder-revert-color: var(--scl-color-black);
  --stars-disabled-color: var(--scl-color-grey-50);
  --stars-spacing: var(--scl-spacing-2);
  --stars-transition: color 0.1s;
  --stars-size: 24px;
}

:host([star-size='small']) {
  --stars-size: 16px;
}

[part='container'] {
  display: inline-flex;
  flex-direction: column;
  position: relative;
}

[part='range-slider'] {
  position: absolute;
  border: 0;
  margin: 0;
  left: calc(-1 * var(--stars-size));
  bottom: 0;
  transform: ;
  z-index: 2;
  width: calc(100% + var(--stars-size) * 2);
  opacity: 0;
  height: var(--stars-size);
}

@media (hover: hover) {
  /* hide the range slider from pointing device */
  [part='range-slider'] {
    clip: rect(0 0 0 0);
    clip-path: inset(50%);
    height: 1px;
    overflow: hidden;
    white-space: nowrap;
    width: 1px;
  }
}

input[type='range']::-webkit-slider-thumb {
  -webkit-appearance: none;
  border: 0;
  height: var(--stars-size);
  width: var(--stars-size);
  border-radius: 50%;
  background: #ffffff;
  cursor: pointer;
  margin-left: 0px;
  margin-top: 0;
}

[part='wrapper'] {
  display: inline-flex;
  border-radius: var(--scl-spacing-4);
  margin-left: calc(-1 * var(--stars-spacing));
}

[part='wrapper']:hover [part='star'] [part='selected-star'] {
  opacity: 1;
}

[part='star'] > * {
  /* prevent event bubbling for icons  */
  pointer-events: none;
}

[part='label'] {
  margin-bottom: var(--scl-spacing-8);
  white-space: nowrap;
}

[part='star'] {
  --decimal-width: 100%;

  display: inline-flex;
  position: relative;
  color: var(--stars-color);
  cursor: pointer;
  padding: 0 var(--stars-spacing);
}

[part='placeholder-star'] {
  display: inline-flex;
  color: var(--stars-placeholder-color);
  transition: var(--stars-transition);
}

[part='selected-star'] {
  color: var(--stars-inactive-color);
  margin-left: 0px;
  position: relative;
  transition: var(--stars-transition);
}

.icon-clip {
  overflow: hidden;
  position: absolute;
  width: var(--decimal-width);
  height: 100%;
}

[part='star'][data-half] {
  --decimal-width: calc(50% - var(--stars-spacing));
}

/* selected state */

[part='star'][data-selected] {
  --stars-placeholder-color: transparent;
}

[part='star'][data-half] {
  --stars-placeholder-color: currentColor;
}

[part='star'][data-selected] [part='selected-star'] {
  color: var(--stars-color);
}

/* hover states */
@media (hover: hover) {
  [part='wrapper']:hover {
    --stars-color: var(--stars-hover-color);
    --stars-placeholder-color: transparent;
  }

  [part='wrapper']:hover [part='star'] {
    --decimal-width: 100%;
  }

  [part='wrapper']:hover [part='star'][data-half] {
    --stars-placeholder-color: transparent;
  }

  [part='wrapper']:hover [part='selected-star'] {
    color: var(--stars-color);
  }

  [part='star']:hover ~ [part='star'] [part='selected-star'] {
    color: var(--stars-inactive-color);
  }

  [part='star']:hover ~ [part='star'] [part='placeholder-star'] {
    --stars-placeholder-color: var(--stars-placeholder-revert-color);
  }
}

/* active states */

[part='wrapper']:active {
  --stars-color: var(--stars-active-color);
}

/* focus states */

:host(:focus-within) [part='wrapper'],
[part='wrapper']:focus {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
  outline: none;
}

/* readonly states */

:host([readonly]) [part='wrapper'] {
  --stars-placeholder-color: var(--stars-disabled-color);

  pointer-events: none;
}

/* disabled states */

:host([disabled]) [part='label'] {
  color: var(--stars-disabled-color);
}

:host([disabled]) [part='wrapper'] {
  --stars-color: var(--stars-disabled-color);
  --stars-placeholder-color: var(--stars-disabled-color);

  pointer-events: none;
}
