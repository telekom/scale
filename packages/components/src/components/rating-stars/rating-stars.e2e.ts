/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { E2EElement, E2EPage, newE2EPage } from '@stencil/core/testing';

describe('rating-stars', () => {
  let page: E2EPage;
  let component: E2EElement;

  beforeEach(async () => {
    page = await newE2EPage();
    await page.setContent(
      `<scale-rating-stars precision="1"></scale-rating-stars>`
    );
    component = await page.find('scale-rating-stars');
  });

  test('renders', async () => {
    expect(component).not.toBeNull();
  });

  test('find div with class rating, which is not null', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    expect(container).not.toBeNull();
  });

  test('click star 3 will return a rating of 3', async () => {
    const star = await page.find('scale-rating-stars >>> #star-3');
    expect(await component.getProperty('value')).toEqual(0);
    await star.click();
    expect(await component.getProperty('value')).toEqual(3);
  });

  test('click star 3 will return a rating of 2.5 (precision="0.5")', async () => {
    page = await newE2EPage();
    await page.setContent(
      `<scale-rating-stars precision="0.5"></scale-rating-stars>`
    );
    component = await page.find('scale-rating-stars');

    const star = await page.find('scale-rating-stars >>> #star-3');
    expect(await component.getProperty('value')).toEqual(0);
    await star.click();
    expect(await component.getProperty('value')).toEqual(2.5);
  });

  test('has no class rating-focus initially', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    expect(container).not.toHaveClass('rating-focus');
  });

  test('gets class rating-focus on press Tab and loses it on click', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    expect(container).toHaveClass('rating-focus');
    await container.click();
    expect(container).not.toHaveClass('rating-focus');
  });

  test('gets class rating-focus on press Tab and keeps it on press Tab', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    expect(container).toHaveClass('rating-focus');
    await container.press('Tab');
    expect(container).toHaveClass('rating-focus');
  });

  test('gets rating of 2 when navigating two to the right with ArrowRight', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    expect(container).toHaveClass('rating-focus');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    expect(await component.getProperty('value')).toEqual(2);
  });

  test('gets rating of 2 when navigating two to the right with ArrowRight (precision="0.5")', async () => {
    page = await newE2EPage();
    await page.setContent(
      `<scale-rating-stars precision="0.5"></scale-rating-stars>`
    );
    component = await page.find('scale-rating-stars');
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    expect(await component.getProperty('value')).toEqual(2);
  });

  test('gets rating of 4 when navigating to star-4 with ArrowRight and leaving with Tab', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    expect(container).toHaveClass('rating-focus');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    await container.press('ArrowRight');
    await container.press('Tab');
    expect(await component.getProperty('value')).toEqual(4);
  });

  test('gets rating of max when pressing End', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    await container.press('End');
    expect(await component.getProperty('value')).toEqual(5);
  });

  test('gets rating of 0 when pressing Home', async () => {
    const container = await page.find('scale-rating-stars >>> .rating');
    await container.press('Tab');
    await container.press('End');
    await container.press('Home');
    expect(await component.getProperty('value')).toEqual(0);
  });
});
