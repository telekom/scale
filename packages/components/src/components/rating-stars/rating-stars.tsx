/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

/*
adapted from shoelace's rating component
https://github.com/shoelace-style/shoelace/blob/next/src/components/rating/rating.ts
*/

import {
  Component,
  h,
  Prop,
  Host,
  Element,
  State,
  Event,
  EventEmitter,
} from '@stencil/core';
import { clamp, handleListeners } from './utils/utils';
import classNames from 'classnames';
import statusNote from '../../utils/status-note';

@Component({
  tag: 'scale-rating-stars',
  styleUrl: 'rating-stars.css',
  shadow: true,
})
export class RatingStars {
  element: HTMLElement;

  @Element() hostElement: HTMLElement;
  /** (optional) hoverValue  */
  @State() hoverValue = 0;
  /** (optional) hoverValue  */
  @State() isHovering = false;
  /** (optional) max  */
  @Prop() max = 5;
  /** (optional) value  */
  @Prop({ mutable: true }) value = 0;
  /** (optional) small  */
  @Prop() small = false;
  /** (optional) disabled  */
  @Prop() disabled = false;
  /** (optional) ariaLabelTranslation  */
  @Prop({ mutable: true })
  ariaLabelTranslation = `${this.value} out of ${this.max} stars`;
  /** (optional) precision  */
  @Prop() precision = 1;
  /** (optional) slider label */
  @Prop() label?: string;
  /** Emitted when the value has changed. */
  @Event() scaleRatingChange!: EventEmitter<{ value: number }>;

  colorFilled = `var(--scl-color-primary)`;
  colorBlank = `var(--scl-color-grey-50)`;
  size = this.small ? '16px' : '24px';

  renderIcon = (color: string, size: string, selected?: boolean) => {
    if (selected) {
      return `<scale-icon-action-favorite color=${color} size=${size} selected />`;
    } else {
      return `<scale-icon-action-favorite color=${color} size=${size} />`;
    }
  };
  connectedCallback() {
    statusNote({ source: this.hostElement, tag: 'beta' });
    this.handleMouseEnter = this.handleMouseEnter.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.handleMouseClick = this.handleMouseClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleTouchStart = this.handleTouchStart.bind(this);
    this.handleTouchEnd = this.handleTouchEnd.bind(this);
    this.handleTouchMove = this.handleTouchMove.bind(this);
  }

  componentDidLoad() {
    handleListeners(this.element, 'addListeners');
  }

  disconnectedCallback() {
    handleListeners(this.element, 'removeListeners');
  }

  handleMouseLeave() {
    this.isHovering = false;
    return false;
  }

  handleMouseEnter() {
    if (!this.disabled) {
      this.isHovering = true;
      return true;
    }
  }

  handleMouseMove(event: MouseEvent) {
    this.hoverValue = this.getValueFromXPosition(null, event);
  }

  handleMouseClick(event: MouseEvent) {
    this.setValue(this.getValueFromXPosition(null, event));
  }

  handleKeyDown(event: KeyboardEvent) {
    if (this.disabled) {
      return;
    }

    if (event.key === 'ArrowRight') {
      const valuePlus = this.value + this.precision;
      this.value = clamp(valuePlus, 0, this.max);
      this.emitRatingValue();
      event.preventDefault();
    }

    if (event.key === 'ArrowLeft') {
      const ratingMinus = this.value - this.precision;
      this.value = clamp(ratingMinus, 0, this.max);
      this.emitRatingValue();
      event.preventDefault();
    }

    if (event.key === 'Home') {
      this.value = 0;
      this.emitRatingValue();
      event.preventDefault();
    }

    if (event.key === 'End') {
      this.value = this.max;
      this.emitRatingValue();
      event.preventDefault();
    }
  }

  getValueFromTouchPosition(event: TouchEvent) {
    return this.getValueFromXPosition(event);
  }

  handleTouchStart(event: TouchEvent) {
    this.hoverValue = this.getValueFromTouchPosition(event);

    // Prevent scrolling when touch is initiated
    event.preventDefault();
  }

  handleTouchMove(event: TouchEvent) {
    this.isHovering = true;
    this.hoverValue = this.getValueFromTouchPosition(event);
  }

  handleTouchEnd(event: TouchEvent) {
    this.isHovering = false;
    this.setValue(this.hoverValue);

    // Prevent click on mobile devices
    event.preventDefault();
  }

  setValue(newValue: number) {
    if (this.disabled) {
      return;
    }

    this.value = newValue === this.value ? 0 : newValue;
    this.emitRatingValue();
    this.isHovering = false;
  }

  getValueFromXPosition(evTou?: TouchEvent, evKey?: MouseEvent) {
    const positionX = evTou ? evTou.touches[0].clientX : evKey.clientX;
    const containerLeft = this.element.getBoundingClientRect().left;
    const containerWidth = this.element.getBoundingClientRect().width;
    return clamp(
      this.roundToPrecision(
        ((positionX - containerLeft) / containerWidth) * this.max,
        this.precision
      ),
      0,
      this.max
    );
  }

  roundToPrecision(numberToRound: number, precision = 1) {
    const multiplier = 1 / precision;
    return Math.ceil(numberToRound * multiplier) / multiplier;
  }

  getAriaLabel() {
    return this.ariaLabelTranslation
      .replace(/\$\{value\}/gi, this.value.toString())
      .replace(/\$\{max\}/gi, this.max.toString());
  }

  emitRatingValue() {
    this.scaleRatingChange.emit({ value: this.value });
  }

  render() {
    const counter = Array.from(Array(this.max).keys());
    const displayValue = this.isHovering ? this.hoverValue : this.value;

    return (
      <Host>
        <div
          class={this.getCssClassMap()}
          ref={(el) => (this.element = el)}
          onMouseMove={this.handleMouseMove}
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}
          onClick={this.handleMouseClick}
          onKeyDown={this.handleKeyDown}
          onTouchStart={this.handleTouchStart}
          onTouchEnd={this.handleTouchEnd}
          onTouchMove={this.handleTouchMove}
          tabIndex={this.disabled ? -1 : 0}
          role="figure"
          aria-describedby="rating__description"
          aria-label={this.label}
        >
          <label class="rating__label" aria-hidden="true">
            {this.label}
          </label>
          <span
            id="rating__description"
            innerHTML={this.getAriaLabel()}
            hidden
          ></span>
          <span class="rating__symbols" aria-hidden="true">
            {counter.map((index) => (
              <span
                class="rating__symbol-wrapper"
                onMouseEnter={this.handleMouseEnter}
              >
                <span
                  role="presentation"
                  style={{
                    clipPath:
                      Math.ceil(displayValue) >= index + 1
                        ? `inset(0 ${
                            (Math.ceil(displayValue) - index) * 100
                          }% 0 0)`
                        : null,
                  }}
                  class={{
                    rating__symbol: true,
                    'rating__symbol--hover':
                      this.isHovering && Math.ceil(displayValue) === index + 1,
                  }}
                  innerHTML={this.renderIcon(this.colorBlank, this.size)}
                  id={`star-${index + 1}`}
                />
              </span>
            ))}
          </span>
          <span
            class="rating__symbols rating__symbols--indicator"
            aria-hidden="true"
          >
            {counter.map((index) => (
              <span
                class="rating__symbol-wrapper"
                onMouseEnter={this.handleMouseEnter}
              >
                <span
                  role="presentation"
                  style={{
                    clipPath:
                      displayValue > index + 1
                        ? null
                        : `inset(0 ${100 - (displayValue - index) * 100}% 0 0)`,
                  }}
                  class={{
                    rating__symbol: true,
                    'rating__symbol--hover':
                      this.isHovering && Math.ceil(displayValue) === index + 1,
                  }}
                  innerHTML={this.renderIcon(this.colorFilled, this.size, true)}
                />
              </span>
            ))}
          </span>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'rating',
      this.disabled && 'rating--disabled',
      this.isHovering && 'rating--hover',
      this.small && 'rating--small',
      this.label && 'rating--has-label'
    );
  }
}
