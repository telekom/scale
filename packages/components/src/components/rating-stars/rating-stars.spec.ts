/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { RatingStars } from './rating-stars';

describe('RatingStars', () => {
  let page: SpecPage;

  describe('props', () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars></scale-rating-stars>`,
      });
    });

    it('check non default props', async () => {
      page.root.maxRating = 7;
      page.root.minRating = 1;
      page.root.rating = 4;
      page.root.starSize = 'small';
      page.root.disabled = true;
      page.root.ariaLabelTranslation = '$value aus $maxValue Sternen';
      page.root.label = 'Rating Label';
      await page.waitForChanges();
      expect(page.rootInstance.maxRating).toBe(7);
      expect(page.rootInstance.minRating).toBe(1);
      expect(page.rootInstance.rating).toBe(4);
      expect(page.rootInstance.starSize).toBe('small');
      expect(page.rootInstance.disabled).toBe(true);
      expect(page.rootInstance.ariaLabelTranslation).toBe(
        '$value aus $maxValue Sternen'
      );
      expect(page.rootInstance.label).toBe('Rating Label');
    });
  });

  describe('snapshots', () => {
    it('rating 2', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars rating="2"></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('rating 3, minRating 1', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars rating="3" min-rating="1"></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('rating 2, maxRating 7', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars rating="2" max-rating="7"></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('rating 3, disabled', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars disabled></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('emitter', () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars></scale-rating-stars>`,
      });
    });

    /* Error Message: TypeError: ev.composedPath is not a function

    it('should change rating mouseUp', async () => {
      expect(page.rootInstance.value).toBe(3);
      const ratingStarTwo = page.root.shadowRoot.querySelector(
        '[data-value="2"]'
      );
      console.log(ratingStarTwo.getAttribute('data-value'));
      await page.waitForChanges();
      ratingStarTwo.dispatchEvent(
        new MouseEvent('mouseup', {
          view: window,
          bubbles: true,
          cancelable: true,
        })
      );
      await page.waitForChanges();
      expect(page.rootInstance.value).toBe(2);
    }); 
*/
  });
});
