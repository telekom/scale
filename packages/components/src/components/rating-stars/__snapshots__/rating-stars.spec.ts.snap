// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingStars snapshots rating 0 should match snapshot 1`] = `
<scale-rating-stars value="0">
  <mock:shadow-root>
    <div aria-describedby="rating__description" class="rating" role="figure" tabindex="0">
      <label aria-hidden="true" class="rating__label"></label>
      <span hidden="" id="rating__description">
        0 out of 5 stars
      </span>
      <span aria-hidden="true" class="rating__symbols">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-1" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-2" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-3" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-4" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-5" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
      <span aria-hidden="true" class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 300% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 400% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 500% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;

exports[`RatingStars snapshots rating 2 and disabled should match snapshot 1`] = `
<scale-rating-stars disabled="" value="2">
  <mock:shadow-root>
    <div aria-describedby="rating__description" class="rating rating--disabled" role="figure" tabindex="-1">
      <label aria-hidden="true" class="rating__label"></label>
      <span hidden="" id="rating__description">
        2 out of 5 stars
      </span>
      <span aria-hidden="true" class="rating__symbols">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-1" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-2" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-3" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-4" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-5" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
      <span aria-hidden="true" class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 0% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 300% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;

exports[`RatingStars snapshots rating 3 should match snapshot 1`] = `
<scale-rating-stars value="3">
  <mock:shadow-root>
    <div aria-describedby="rating__description" class="rating" role="figure" tabindex="0">
      <label aria-hidden="true" class="rating__label"></label>
      <span hidden="" id="rating__description">
        3 out of 5 stars
      </span>
      <span aria-hidden="true" class="rating__symbols">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-1" role="presentation" style="clip-path: inset(0 300% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-2" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-3" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-4" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-5" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
      <span aria-hidden="true" class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 0% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="24px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;

exports[`RatingStars snapshots rating 4 and small should match snapshot 1`] = `
<scale-rating-stars small="" value="4">
  <mock:shadow-root>
    <div aria-describedby="rating__description" class="rating rating--small" role="figure" tabindex="0">
      <label aria-hidden="true" class="rating__label"></label>
      <span hidden="" id="rating__description">
        4 out of 5 stars
      </span>
      <span aria-hidden="true" class="rating__symbols">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-1" role="presentation" style="clip-path: inset(0 400% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-2" role="presentation" style="clip-path: inset(0 300% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-3" role="presentation" style="clip-path: inset(0 200% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-4" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" id="star-5" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-grey-50)" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
      <span aria-hidden="true" class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 0% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
        <span class="rating__symbol-wrapper">
          <span class="rating__symbol" role="presentation" style="clip-path: inset(0 100% 0 0);">
            <scale-icon-action-favorite color="var(--scl-color-primary)" selected="" size="16px"></scale-icon-action-favorite>
          </span>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;
